### DISCUSSION TITLE
Reusable task class via decorator that adds common utility logic in run

### DISCUSSION THREAD

**Discussion by thenewguy** [celery/celery#7026] [Q&A]
I am not sure if this issue is a feature request or a request for guidance.  I do not see how this could be accomplished, so I guess it is a feature request -- but perhaps it is already possible?

I am trying to swap from using class based tasks to decorator tasks.  Simple tasks are easy to swap over.  However, some of my tasks are based around this class based task: https://github.com/PolicyStat/jobtastic/blob/v2.1.1/jobtastic/task.py#L313

The Jobtastic Task Class provides a run method that prevents duplicate tasks from running simultaneously and abstracts some cruft that is pretty handy.  Then its run method calls `calculate_result()` to perform the actual work of the task you create.

What has been holding this back is since the [decorator makes your function the run method](https://docs.celeryproject.org/en/stable/userguide/tasks.html#custom-task-classes), how can I implement this functionality without rewriting it in every task?  If the run() method of the Jobtastic Task is overridden, the utility of the task is lost.

Is there a way to incorporate the advanced runtime logic of the Jobtastic Task?  Is it possible to tell the decorator what method the wrapped function should become?  Or would an additional `run()` wrapper be required to handle this with a decorated approach?

**Comment by georgepsarakis** [comment]
If I understand correctly what you are trying to achieve, I am not sure there is a fully automated way to provide such an API. To summarize, you want to define other tasks using the `@app.task` decorator, with the `base` parameter being the Jobtastic Task class? If the new task's implementation is [calculate_result](https://github.com/PolicyStat/jobtastic/blob/9838a6caa3025bbc379708617b88d5858cdc2dfc/jobtastic/task.py#L342), you could probably accomplish it by defining a context manager in the base class and require its use in each task.

How does that sound?

**Comment by thenewguy** [comment]
@georgepsarakis Yes - using a reusable base class that implements similar functionality so that we can move to the recommended way of calling tasks

I am not visualizing what you mean though.  Are you saying to move the logic handled by the Jobtastic base class into a context manager and using the context block as the replacement for the "calculate_result" method?
